File: common/src/main/java/com/hartwig/actin/util/json/GsonSerializer.java
package com.hartwig.actin.util.json;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Set;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import org.jetbrains.annotations.NotNull;

public final class GsonSerializer {

    private GsonSerializer() {
    }

    @NotNull
    public static Gson create() {
        // If we don't register an explicit type adapter for LocalDate, GSON using reflection internally to create serialize these objects
        return new GsonBuilder().serializeNulls()
                .enableComplexMapKeySerialization().registerTypeAdapter(LocalDateTime.class, new GsonLocalDateTimeAdapter())
                .registerTypeAdapter(LocalDate.class, new GsonLocalDateAdapter())
                .registerTypeHierarchyAdapter(Set.class, new GsonSetAdapter<>())
                .create();
    }
}

File: common/src/main/kotlin/com/hartwig/actin/util/json/GsonSerializer.kt
package com.hartwig.actin.util.json

import com.google.gson.GsonBuilder
import java.time.LocalDate
import java.time.LocalDateTime

object GsonSerializer {

    fun createBuilder(): GsonBuilder = GsonBuilder().serializeNulls()
        .enableComplexMapKeySerialization()
        .serializeSpecialFloatingPointValues()
        .registerTypeAdapter(LocalDateTime::class.java, GsonLocalDateTimeAdapter())
        .registerTypeAdapter(LocalDate::class.java, GsonLocalDateAdapter())
        .registerTypeHierarchyAdapter(Set::class.java, GsonSetAdapter<Any>())
}